src/elements/element.rectangle.js

```diff
@@ -50,11 +50,106 @@ module.exports = function(Chart) {
 	}
 
 	Chart.elements.Rectangle = Chart.Element.extend({
-		draw: function() {
+		getBorderWidth: function(top, right, bottom, left) {
+			// Get calculated border width
+			var vm = this._view;
+			// borderWidth should be less than bar width and bar height.
+			// ..,maximum border width for left-right borders
+			var barSizeLeftRight = Math.abs(left - right);
+			// ...maximum border width for top-bottom borders
+			var barSizeTopBottom = Math.abs(top - bottom);
+
+			function ceiling(val, max) {
+				return val > max ? max : val;
+			}
+
+			// Like CSS:
+			// - one border to rule them all
+			// - top-bottom + left-right
+			// - top + right + bottom + left
+			return !vm.borderWidth ? [] :
+				vm.borderWidth instanceof Array && vm.borderWidth.length === 1 ? {
+					top: ceiling(vm.borderWidth[0], barSizeTopBottom),
+					right: ceiling(vm.borderWidth[0], barSizeLeftRight),
+					bottom: ceiling(vm.borderWidth[0], barSizeTopBottom),
+					left: ceiling(vm.borderWidth[0], barSizeLeftRight)
+				} :
+				vm.borderWidth instanceof Array && vm.borderWidth.length === 2 ? {
+					top: ceiling(vm.borderWidth[0], barSizeTopBottom),
+					right: ceiling(vm.borderWidth[1], barSizeLeftRight),
+					bottom: ceiling(vm.borderWidth[0], barSizeTopBottom),
+					left: ceiling(vm.borderWidth[1], barSizeLeftRight)
+				} :
+				vm.borderWidth instanceof Array ? {
+					top: ceiling(vm.borderWidth[0], barSizeTopBottom),
+					right: ceiling(vm.borderWidth[1], barSizeLeftRight),
+					bottom: ceiling(vm.borderWidth[2], barSizeTopBottom),
+					left: ceiling(vm.borderWidth[3], barSizeLeftRight)
+				} :
+				{
+					top: ceiling(vm.borderWidth, barSizeTopBottom),
+					right: ceiling(vm.borderWidth, barSizeLeftRight),
+					bottom: ceiling(vm.borderWidth, barSizeTopBottom),
+					left: ceiling(vm.borderWidth, barSizeLeftRight)
+				};
+		},
+		drawShape: function(borders) {
+			// Draw the inside shape...
 			var ctx = this._chart.ctx;
 			var vm = this._view;
-			var left, right, top, bottom, signX, signY, borderSkipped;
-			var borderWidth = vm.borderWidth;
+
+			ctx.beginPath();
+			ctx.fillStyle = vm.backgroundColor;
+			ctx.strokeStyle = 'rgba(0,0,0,0)';
+			ctx.lineWidth = 0;
+			ctx.moveTo(borders.bottom[0][0], borders.bottom[0][1]);
+			for (var i in borders) {
+				if (borders.hasOwnProperty(i)) {
```

Comment:
Seems counter efficient to iterate on this `borders` object and have all these `hasOwnProperty` useless calls. Since it's an internal variable, why not using an array instead and convert this `for..in` loop to a tradition `for(var i=0; i<4; ++i)` loop?